OpenRA.Mods.Common/UpdateRules/Rules/20210321/UnhardcodeSquadManager.cs

```diff
@@ -11,33 +11,128 @@
+			var aircraft = new List<string>();
+			var vips = new List<string>();
+
+			foreach (var actor in resolvedActors)
+			{
+				if (actor.Key.StartsWith('^'))
+					continue;
+
+				var isVip = false;
+				var isBuildable = false;
+				var isBuilding = false;
+				var isAircraft = false;
+				var isExcluded = false;
+				var canAttack = false;
+				var isKillable = false;
+
+				foreach (var trait in actor.Value.Nodes)
+				{
+					if (trait.IsRemoval())
+						continue;
+
+					if (trait.Key == "Buildable")
+					{
+						isBuildable = true;
+						continue;
+					}
+
+					if (trait.Key == "Aircraft")
+					{
+						isAircraft = true;
+						continue;
+					}
+
+					if (trait.Key == "Health")
+					{
+						isKillable = true;
+						continue;
+					}
+
+					if (vipsNames.Any(v => v == trait.Key))
+					{
+						isVip = true;
+						continue;
+					}
+
+					if (buildings.Any(v => v == trait.Key))
+					{
+						isBuilding = true;
+						continue;
+					}
+
+					if (excludedBuildings.Any(v => v == trait.Key))
+					{
+						isExcluded = true;
+						continue;
+					}
+
+					if (attackBase.Any(ab => ab == trait.Key))
+						canAttack = true;
+				}
 
-			var aircraft = modData.DefaultRules.Actors.Values.Where(a => a.HasTraitInfo<AircraftInfo>() && a.HasTraitInfo<AttackBaseInfo>()).Select(a => a.Name);
-			var airUnits = new MiniYamlNode("AirUnitsTypes", FieldSaver.FormatValue(aircraft.ToList()));
-			addNodes.Add(airUnits);
+				if (isAircraft && isBuildable && canAttack && isKillable)
+				{
+					var name = actor.Key.ToLower();
+					if (!aircraft.Contains(name))
+						aircraft.Add(name);
+				}
 
-			var vips = modData.DefaultRules.Actors.Values.Where(a => a.HasTraitInfo<HarvesterInfo>() || a.HasTraitInfo<BaseBuildingInfo>() || (a.HasTraitInfo<BuildingInfo>() && a.HasTraitInfo<BuildableInfo>() && !a.HasTraitInfo<LineBuildInfo>() && !a.HasTraitInfo<PlugInfo>())).Select(a => a.Name);
-			var protection = new MiniYamlNode("ProtectionTypes", FieldSaver.FormatValue(vips.ToList()));
-			addNodes.Add(protection);
+				if (isBuildable && isKillable && (isVip || (isBuilding && !isExcluded)))
+				{
+					var name = actor.Key.ToLower();
+					if (!vips.Contains(name))
+						vips.Add(name);
+				}
+			}
 
+			addNodes.Add(new MiniYamlNode("AirUnitsTypes", FieldSaver.FormatValue(aircraft)));
+			addNodes.Add(new MiniYamlNode("ProtectionTypes", FieldSaver.FormatValue(vips)));
+
+			yield break;
+		}
+
+		bool anyAdded = false;
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
 			foreach (var squadManager in actorNode.ChildrenMatching("SquadManagerBotModule"))
+			{
 				foreach (var addNode in addNodes)
 					squadManager.AddNode(addNode);
 
+				anyAdded = true;
+			}
+
 			yield break;
 		}
+
+		public override IEnumerable<string> AfterUpdate(ModData modData)
+		{
+			if (anyAdded)
+				yield return "New fields were added to SquadManagerBotModule: `AirUnitsTypes` and `ProtectionTypes`.";
```

Comment:
These don't need to be reported if they've already been set automatically.

Reply:
I've been asked to have them reported

Reply:
I asked for that because we can't guarantee correctness, so it's best if modders verify manually.

Reply:
I just realised this doesn't actually tell the user to check what was updated or what values were added.

Reply:
You mean "actors that were added to those" as this is a list of actors on a trait on the player 😜
Either that or just a message saying "please verify the automated changes are correct" would be nice, thanks.

Reply:
I updated the comment.